Note: This has been taken from https://engineering.ucsb.edu/~shell/che210d/numpy.pdf
(a list of useful numpy apis)

In [9]:
import numpy as np

### Array mathematics

In [10]:
a = np.array([1,2,3], float)
b = np.array([5,2,6], float)

In [11]:
a + b

array([ 6.,  4.,  9.])

In [12]:
a/b

array([ 0.2,  1. ,  0.5])

In [13]:
a % b

array([ 1.,  0.,  3.])

In [14]:
b % a

array([ 0.,  0.,  0.])

In [15]:
b**a #power

array([   5.,    4.,  216.])

In [16]:
a = np.array([1,2,3], float)
b = np.array([4,5], float)
#a + b

In [17]:
a = np.array([[1, 2], [3, 4], [5, 6]], float)
b = np.array([1, 100], float)
a + b

array([[   2.,  102.],
       [   4.,  104.],
       [   6.,  106.]])

In [18]:
np.array([-1, 3], float)

array([-1.,  3.])

In [19]:
b = np.array([-1., 3.], float)
b

array([-1.,  3.])

In [20]:
 b[:,np.newaxis]

array([[-1.],
       [ 3.]])

In [21]:
 b[:,]

array([-1.,  3.])

In [22]:
b[:,]

array([-1.,  3.])

In [23]:
x = b[:,np.newaxis]
x

array([[-1.],
       [ 3.]])

In [24]:
x.shape

(2, 1)

In [25]:
 b[np.newaxis,:]

array([[-1.,  3.]])

In [26]:
arr = np.arange(4)
print arr.shape
arr

(4,)


array([0, 1, 2, 3])

In [27]:
arr = np.arange(4)
arr[np.newaxis,:]

array([[0, 1, 2, 3]])

In [28]:
col_vec = arr[:, np.newaxis]

In [29]:
col_vec.shape

(4, 1)

In [30]:
a = np.array([[0, 2], [3, -1], [3, 5]], float)
a

array([[ 0.,  2.],
       [ 3., -1.],
       [ 3.,  5.]])

In [31]:
for (x, y) in a:
    print x+y

2.0
2.0
8.0


In [32]:
#equivalent but much much faster
a.sum(axis=1)

array([ 2.,  2.,  8.])

### Basic array operations

In [33]:
a = np.array([2, 4, 3], float)
print a.sum()
print a.prod()

9.0
24.0


In [34]:
a.std()

0.81649658092772603

In [35]:
np.sqrt(((a - a.mean())**2).sum()/a.size)

0.81649658092772603

In [36]:
#The argmin and argmax functions return the array indices of the minimum and maximum values:
a = np.array([10, 100, 0, 9999], float)
print a.argmax()
print a.argmin()

3
2


In [37]:
a = np.array([[0, 200], [3, -1], [3, 5]], float)
a

array([[   0.,  200.],
       [   3.,   -1.],
       [   3.,    5.]])

In [38]:
a.mean(axis=0)

array([  2.,  68.])

In [39]:
a.mean(axis=1)

array([ 100.,    1.,    4.])

In [40]:
a.min(axis=1)

array([ 0., -1.,  3.])

In [41]:
a.max(axis=0)

array([   3.,  200.])

In [42]:
a = np.array([6, 2, 5, -1, 0], float)

In [43]:
sorted(a)

[-1.0, 0.0, 2.0, 5.0, 6.0]

In [44]:
a

array([ 6.,  2.,  5., -1.,  0.])

#### Values in an array can be "clipped" to be within a prespecified range. This is the same as
applying min(max(x, minval), maxval) to each element x in an array. Like winsorization

In [45]:
a = np.array([6, 2, 5, -1, 0], float)
a.clip(0,5)

array([ 5.,  2.,  5.,  0.,  0.])

In [46]:
a = np.array([1, 1, 4, 5, 5, 5, 7], float)
np.unique(a)

array([ 1.,  4.,  5.,  7.])

In [47]:
a = np.array([[1, 2], [3, 4]], float)
print(a)

[[ 1.  2.]
 [ 3.  4.]]


In [48]:
a.diagonal()

array([ 1.,  4.])

In [49]:
a = np.array([1, 3, 0], float)
b = np.array([0, 3, 2], float)
a > b

array([ True, False, False], dtype=bool)

In [50]:
a < b

array([False, False,  True], dtype=bool)

In [51]:
a == b

array([False,  True, False], dtype=bool)

In [52]:
 np.logical_and(a > b, a < b)

array([False, False, False], dtype=bool)

In [53]:
np.logical_or(np.logical_or(a > b, b > a), a == b)

array([ True,  True,  True], dtype=bool)

In [54]:
a = np.array([1, 3, 0], float)
a > 2

array([False,  True, False], dtype=bool)

#### The any and all operators can be used to determine whether or not any or all elements of a
Boolean array are true:

In [55]:
c = np.array([ True, False, False], bool)
any(c)

True

In [56]:
all(c)

False

The *** where ***  function forms a new array from two arrays of equivalent size using a Boolean filter to choose between elements of the two. Its basic syntax is where(boolarray,
truearray, falsearray):

In [57]:
a = np.array([1, 3, 0], float)
np.where(a != 0.0, 1.0/a, a)

/Users/jvsingh/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:2: RuntimeWarning: divide by zero encountered in divide
  from ipykernel import kernelapp as app


array([ 1.        ,  0.33333333,  0.        ])

In [58]:
np.where(a > 0, 3, 2)

array([3, 3, 2])

A number of functions allow testing of the values in an array. The *** nonzero *** function gives a tuple of indices of the nonzero values in an array. The number of items in the tuple equals the number of axes of the array:

In [59]:
a = np.array([[10, 0], [0, 5]], float)
print a
a.nonzero()

[[ 10.   0.]
 [  0.   5.]]


(array([0, 1]), array([0, 1]))

In [60]:
a = np.array([[10, 1], [3, 0]], float)
print a
a.nonzero()

[[ 10.   1.]
 [  3.   0.]]


(array([0, 0, 1]), array([0, 1, 0]))

#### It is also possible to test whether or not values are NaN ("not a number") or finite:

In [61]:
a = np.array([1, np.NaN, np.Inf], float)
a

array([  1.,  nan,  inf])

In [62]:
np.isnan(a)

array([False,  True, False], dtype=bool)

In [63]:
np.isfinite(a)

array([ True, False, False], dtype=bool)

#### Array item selection and manipulation

In [64]:
a = np.array([[6, 4], [5, 9]], float)
print a
a >= 6
print a >= 6

[[ 6.  4.]
 [ 5.  9.]]
[[ True False]
 [False  True]]


In [65]:
sel = (a >= 6)
a[sel]

array([ 6.,  9.])

In [66]:
sel = a > 0
a[sel]

array([ 6.,  4.,  5.,  9.])

In [67]:
a[np.logical_and(a > 5, a < 9)]

array([ 6.])

##### Selection using integer arrays

In [68]:
a = np.array([2, 4, 6, 8], float)
b = np.array([0, 0, 1, 3, 2, 1], int)
a[b]

array([ 2.,  2.,  4.,  8.,  6.,  4.])

##### Lists can also be used as selection arrays:

In [69]:
a = np.array([2, 4, 6, 8], float)
a[[0, 0, 1, 3, 2, 1]]

array([ 2.,  2.,  4.,  8.,  6.,  4.])

#### Multidimensional selection

In [70]:
a = np.array([[1, 4], [9, 16]], float)
print a

[[  1.   4.]
 [  9.  16.]]


In [71]:
b = np.array([0, 0, 1, 1, 0], int)
c = np.array([0, 1, 1, 1, 1], int)
a[b,c] #first index for first array and second from second array

array([  1.,   4.,  16.,  16.,   4.])

In [72]:
# ** take *** function
a = np.array([2, 4, 6, 8], float)
b = np.array([0, 0, 1, 3, 2, 1], int)
a.take(b)

array([ 2.,  2.,  4.,  8.,  6.,  4.])

In [73]:
a = np.array([[0, 1], [2, 3]], float)
a

array([[ 0.,  1.],
       [ 2.,  3.]])

In [74]:
b = np.array([0, 0, 1], int)

In [75]:
a.take(b, axis=0)

array([[ 0.,  1.],
       [ 0.,  1.],
       [ 2.,  3.]])

In [76]:
a.take(b, axis=1)

array([[ 0.,  0.,  1.],
       [ 2.,  2.,  3.]])

*** put *** function (opposite of take)

In [77]:
a = np.array([0, 1, 2, 3, 4, 5], float)
b = np.array([9, 8, 7], float)
a.put([0, 3], b)
a

array([ 9.,  1.,  2.,  8.,  4.,  5.])

#### Vector and matrix mathematics

In [78]:
a = np.array([1, 4, 0], float)
b = np.array([2, 2, 1], float)
np.dot(a,b)

10.0

In [79]:
np.outer(a, b)

array([[ 2.,  2.,  1.],
       [ 8.,  8.,  4.],
       [ 0.,  0.,  0.]])

In [80]:
np.inner(a, b)

10.0

In [81]:
np.cross(a, b)

array([ 4., -1., -6.])

#### Linear Algebra calculations

In [82]:
a = np.array([[4, 2, 0], [9, 3, 7], [1, 2, 1]], float)
print a
np.linalg.det(a)

[[ 4.  2.  0.]
 [ 9.  3.  7.]
 [ 1.  2.  1.]]


-48.000000000000028

In [83]:
np.linalg.det(np.array([[4, 2], [9, 3]], float))

-6.0000000000000027

In [84]:
# eigenvalues and eigenvectors 

np.linalg.eig(a)

(array([ 8.85591316,  1.9391628 , -2.79507597]),
 array([[-0.3663565 , -0.54736745,  0.25928158],
        [-0.88949768,  0.5640176 , -0.88091903],
        [-0.27308752,  0.61828231,  0.39592263]]))

In [85]:
vals, vecs= np.linalg.eig(a)
print vecs
print vals

[[-0.3663565  -0.54736745  0.25928158]
 [-0.88949768  0.5640176  -0.88091903]
 [-0.27308752  0.61828231  0.39592263]]
[ 8.85591316  1.9391628  -2.79507597]


In [86]:
x_1 = vecs[:,0] #first eigenvector
lambda_1 = vals[0] #first eigenvalue
np.dot(a, x_1) - lambda_1 * x_1  

array([ -2.66453526e-15,  -4.44089210e-15,  -1.33226763e-15])

In [87]:
#should be close to 0,0,0
np.dot(a, vecs[:,1]) - vals[1] * vecs[:,1]

array([ -1.11022302e-15,  -3.77475828e-15,   0.00000000e+00])

In [88]:
#should be close to 0,0,0
np.dot(a, vecs[:,2]) - vals[2] * vecs[:,2]

array([  0.00000000e+00,   4.44089210e-16,   0.00000000e+00])

In [89]:
b = np.linalg.inv(a)

In [90]:
b

array([[ 0.22916667,  0.04166667, -0.29166667],
       [ 0.04166667, -0.08333333,  0.58333333],
       [-0.3125    ,  0.125     ,  0.125     ]])

In [91]:
np.dot(a, b)

array([[  1.00000000e+00,   5.55111512e-17,   0.00000000e+00],
       [  0.00000000e+00,   1.00000000e+00,   2.22044605e-16],
       [  0.00000000e+00,   1.38777878e-17,   1.00000000e+00]])

In [92]:
a = np.array([[1, 3, 4], [5, 2, 3]], float)

In [93]:
U, s, Vh = np.linalg.svd(a)

In [94]:
U

array([[-0.6113829 , -0.79133492],
       [-0.79133492,  0.6113829 ]])

In [95]:
s

array([ 7.46791327,  2.86884495])

#### Polynomial mathematics

##### Polynomial coefficients from set of roots

In [96]:
np.poly([-1, 1, 1, 10])

array([  1., -11.,   9.,  11., -10.])

so polynomial is $ x^4 - 11 x^3 + 9 x^2 + 11x - 10 $

In [97]:
#roots from polynomial
np.roots([1, 4, -2, 3])

array([-4.57974010+0.j        ,  0.28987005+0.75566815j,
        0.28987005-0.75566815j])

Polynomial integration:

$x^3  + x^2 + x + 1$

In [98]:
 np.polyint([1, 1, 1, 1])

array([ 0.25      ,  0.33333333,  0.5       ,  1.        ,  0.        ])

means $ 0.25 x^4 + 0.333 x^3  + 0.5 x^2 + x + C $ where C is set 0

There are other routines for plynomials for polyadd, polysub, polymul, and polydiv. Also to evalutae plynomial at a given point you may use ** polyval **

the *** polyfit *** function can be used to fit a polynomial of specified order to a set of data using a least-squares approach:

In [99]:
x = [1, 2, 3, 4, 5, 6, 7, 8]
y = [0, 2, 1, 3, 7, 10, 11, 19]
np.polyfit(x, y, 2) #returns the plynomial coefficients

array([ 0.375     , -0.88690476,  1.05357143])

### Statistics functions

In [100]:
a = np.array([1, 4, 3, 8, 9, 2, 3], float)
np.median(a)

3.0

In [101]:
a = np.array([1, 4, 3, 8, 9, 2, 3], float)
np.median(a)

3.0

In [102]:
a = np.array([[1, 2, 1, 3], [5, 3, 1, 8]], float)
np.corrcoef(a)

array([[ 1.        ,  0.72870505],
       [ 0.72870505,  1.        ]])

In [103]:
np.cov(a)

array([[ 0.91666667,  2.08333333],
       [ 2.08333333,  8.91666667]])

In [104]:
2.08333333/np.sqrt(0.91666667 * 8.91666667)

0.72870504403430825

### Random numbers 

There is a submodule called *** random *** . NumPy uses a particular algorithm called the Mersenne Twister to generate
pseudorandom numbers

In [109]:
 np.random.seed(293423)

In [110]:
 np.random.rand(5)

array([ 0.33677247,  0.52693437,  0.79529578,  0.78867702,  0.02147624])

The rand function can be used to generate two-dimensional random arrays, or the resize
function could be employed here:

In [111]:
np.random.rand(2,3)

array([[ 0.84612516,  0.0704939 ,  0.1526965 ],
       [ 0.77831701,  0.80821151,  0.82198398]])

In [112]:
np.random.rand(6).reshape((2,3))

array([[ 0.90239653,  0.8385685 ,  0.02638565],
       [ 0.33681448,  0.46480928,  0.61686496]])

In [113]:
#single 
np.random.random()

0.43767262538051455

In [114]:
np.random.randint(5, 10)

5

In [120]:
np.random.poisson(6.0) #lambda = 6

4

In [121]:
np.random.normal(1.5, 4.0) # mean = 1.5, sd = 4.0

3.6448077733134396

In [122]:
np.random.normal()

0.28542997954753085

In [123]:
np.random.normal(size=5)

array([-0.01270274, -0.74276215, -0.48822038, -0.63834433, -0.63432301])

In [138]:
l = range(10)
print l
np.random.shuffle(l)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


In [139]:
#modifies list inplace
l = [1,5,"TT", 7, "10"]
np.random.shuffle(l)
print l

['10', 1, 5, 'TT', 7]
